### PR TITLE
Feature/refactor peers

### PR DESCRIPTION
**Pull request by bogdan34**

### CODE REVIEW

File: core/controller/peer.ts [hunk 1]
```diff
@@ -1,56 +1,92 @@
...
+        const peers = shuffledPeers.slice(0, config.CONSTANTS.MAX_PEERS_CONNECT_TO - PeerNetworkRepository.count);
+        PeerService.connectPeers(peers);
     }
 
-    @ON('REQUEST_PEERS')
-    sendPeers(action: { peer: Peer, requestId: string }): void {
-        const { peer, requestId } = action;
-        SyncService.sendPeers(peer, requestId);
+    @ON(ActionTypes.REQUEST_PEERS)
+    sendPeers(action: { peerAddress: PeerAddress, requestId: string }): void {
```
Comment (macaqueee) [inline core/controller/peer.ts]: create `type` for this argument

File: core/controller/peer.ts [hunk 2]
```diff
@@ -1,56 +1,92 @@
...
         }
+
         const peers = shuffledPeers.slice(0, config.CONSTANTS.PEERS_DISCOVER.MAX);
-        PeerRepository.connectPeers(peers);
+        PeerService.connectPeers(peers);
     }
+
+    @ON(ActionTypes.HEADERS_RECEIVE)
+    headersReceive(
+        data: {
```
Comment (macaqueee) [inline core/controller/peer.ts]: create `type`

File: core/controller/sync.ts [hunk 3]
```diff
@@ -119,21 +127,23 @@ export class SyncController extends BaseController {
         logger.info(`${LOG_PREFIX}[startSyncBlocks] SYNCHRONIZATION DONE SUCCESS`);
     }
 
-    @ON('REQUEST_BLOCKS')
-    sendBlocks(action: { data: { height: number, limit: number }, peer: Peer, requestId: string }): void {
-        const { data, peer, requestId } = action;
-        SyncService.sendBlocks(data, peer, requestId);
+    @ON(ActionTypes.REQUEST_BLOCKS)
+    sendBlocks(action: { data: { height: number, limit: number }, peerAddress: PeerAddress, requestId: string }): void {
```
Comment (macaqueee) [inline core/controller/sync.ts]: create `type`

File: core/controller/sync.ts [hunk 4]
```diff
@@ -119,21 +127,23 @@ export class SyncController extends BaseController {
...
+        const { data, peerAddress, requestId } = action;
+        SyncService.sendBlocks(data, peerAddress, requestId);
     }
 
-    @ON('PEER_HEADERS_UPDATE')
-    updatePeer(action: { data, peer }): void {
-        const { data, peer } = action;
-        PeerService.update(data, peer);
+    @ON(ActionTypes.PEER_HEADERS_UPDATE)
+    updatePeer(action: { data: Headers, peerAddress: PeerAddress }): void {
```
Comment (macaqueee) [inline core/controller/sync.ts]: create `type`

File: core/driver/socket.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            logger.warn(`[SOCKET][onServerConnect] trying to connect from black ip`);
+            socket.disconnect(true);
+        }
+
+        if (Socket.ioServer.clients().length > config.CONSTANTS.MAX_PEERS_CONNECTED) {
+            logger.warn(`[SOCKET][onServerConnect] too many connections, sorry ${ip}`);
+            socket.disconnect(true);
+        }
+
+        socket.on(SOCKET_EVENTS.HEADERS, (response: string) => {
```
Comment (macaqueee) [inline core/driver/socket.ts]: split this code by extracting callback to some additional function.

File: core/service/peer.ts [hunk 6]
```diff
@@ -1,12 +1,88 @@
...
+import PeerMemoryRepository from 'core/repository/peer/peerMemory';
+import PeerNetworkRepository from 'core/repository/peer/peerNetwork';
+import { FullHeaders, Headers, PeerAddress } from 'shared/model/types';
+import { NetworkPeer } from 'shared/model/networkPeer';
+import SocketDriver, { SOCKET_TYPE } from 'core/driver/socket';
+import SystemRepository from 'core/repository/system';
+import config from 'shared/config';
+import { sortFunction } from 'core/util/socket';
+import { logger } from 'shared/util/logger';
+const LOG_PREFIX = `[Service]][Peer]`;
```
Comment (axelsheva) [inline core/service/peer.ts]: ]]

File: core/repository/peer/peerMemory.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import { FullHeaders, PeerAddress } from 'shared/model/types';
+import { MemoryPeer } from 'shared/model/memoryPeer';
+import { SOCKET_TYPE } from 'core/driver/socket';
+
+class PeerMemoryRepo {
```
Comment (macaqueee) [inline core/repository/peer/peerMemory.ts]: rename to `PeerMemoryRepository`

File: core/repository/peer/peerNetwork.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import { NetworkPeer } from 'shared/model/networkPeer';
+import { PeerAddress } from 'shared/model/types';
+
+
+class PeerNetworkRepo {
```
Comment (macaqueee) [inline core/repository/peer/peerNetwork.ts]: rename to `PeerNetworkRepository`

File: core/service/peer.ts [hunk 9]
```diff
@@ -1,12 +1,88 @@
...
 
-    update(headers, peer) {
-        PeerRepository.peerUpdate(headers, peer);
+        if (PeerNetworkRepository.isBanned(peerAddress) || config.CORE.PEERS.BLACKLIST.indexOf(peerAddress.ip) !== -1) {
+            return false;
+        }
+
+        if (PeerMemoryRepository.has(peerAddress)) {
+            logger.debug(`${LOG_PREFIX}[add] sockets conflict detected on ${peerAddress.ip}`);
+            const peerNetwork = PeerNetworkRepository.get(peerAddress);
```
Comment (macaqueee) [inline core/service/peer.ts]: remove to `networkPeer`

File: core/service/peer.ts [hunk 10]
```diff
@@ -1,12 +1,88 @@
...
+        } else {
+            PeerMemoryRepository.add(peerAddress, headers);
+            PeerNetworkRepository.add(peerAddress, socket);
+        }
+        return true;
+
+    }
+
+    ban(peerAddress: PeerAddress): void {
+        PeerNetworkRepository.addToBan(peerAddress);
```
Comment (macaqueee) [inline core/service/peer.ts]: rename to `addToBanList`

File: core/service/sync.ts [hunk 11]
```diff
@@ -16,72 +13,111 @@ import SharedTransactionRepo from 'shared/repository/transaction';
...
+import PeerNetworkRepository from 'core/repository/peer/peerNetwork';
+import PeerService, { ERROR_NOT_ENOUGH_PEERS } from 'core/service/peer';
+import { ActionTypes } from 'core/util/actionTypes';
+import { BlockData, PeerAddress } from 'shared/model/types';
+import { getRandom, getRandomElements } from 'core/util/common';
+import { NetworkPeer } from 'shared/model/networkPeer';
+import { MemoryPeer } from 'shared/model/memoryPeer';
 
 //  TODO get from env
 const MIN_CONSENSUS = 51;
```
Comment (macaqueee) [inline core/service/sync.ts]: use this variable from config.

File: core/service/sync.ts [hunk 12]
```diff
@@ -16,72 +13,111 @@ import SharedTransactionRepo from 'shared/repository/transaction';
...
+        peer.responseRPC(ActionTypes.RESPONSE_PEERS, peerAddresses, requestId);
     }
 
     sendNewBlock(block: Block): void {
         block.relay += 1;
         if (block.relay < config.CONSTANTS.TRANSFER.MAX_RELAY) {
-            SyncRepository.sendNewBlock(block);
+            const serializedBlock: Block & { transactions: any } = block.getCopy();
+            serializedBlock.transactions = block.transactions.map(trs => SharedTransactionRepo.serialize(trs));
+            PeerService.broadcast(ActionTypes.BLOCK_RECEIVE, { block: serializedBlock });
```
Comment (macaqueee) [inline core/service/sync.ts]: check block for method `getCopy`

File: core/service/sync.ts [hunk 13]
```diff
@@ -98,36 +134,43 @@ export class SyncService implements ISyncService {
...
 
-    async rollback() {
+    async rollback(): Promise<ResponseEntity<any>> {
+        // TODO Return response entity
         const blockSlot = SlotService.getSlotNumber(BlockRepository.getLastBlock().createdAt);
         const prevRound = RoundRepository.getPrevRound();
         if (!prevRound) {
             await BlockService.deleteLastBlock();
-            return;
+            return new ResponseEntity({});
```
Comment (macaqueee) [inline core/service/sync.ts]: remove empty object from arguments

File: core/service/sync.ts [hunk 14]
```diff
@@ -98,36 +134,43 @@ export class SyncService implements ISyncService {
...
+        const peerNetwork = PeerNetworkRepository.get(peerAddress);
+        return await peerNetwork.requestRPC(ActionTypes.REQUEST_BLOCKS, {
             height: lastBlock.height,
             limit: config.CONSTANTS.TRANSFER.REQUEST_BLOCK_LIMIT
-        }, peer);
+        });
     }
 
-    sendBlocks(data: { height: number, limit: number }, peer, requestId): void {
+    sendBlocks(data: { height: number, limit: number }, peerAddress, requestId): void {
```
Comment (macaqueee) [inline core/service/sync.ts]: `type`

File: core/service/sync.ts [hunk 15]
```diff
@@ -144,37 +187,45 @@ export class SyncService implements ISyncService {
         return new ResponseEntity();
     }
 
-    checkCommonBlocks(block: { id: string, height: number }, peer, requestId): void {
+    checkCommonBlocks(block: { id: string, height: number }, peerAddress: PeerAddress, requestId): void {
```
Comment (macaqueee) [inline core/service/sync.ts]: `type`

File: core/service/sync.ts [hunk 16]
```diff
@@ -144,37 +187,45 @@ export class SyncService implements ISyncService {
...
         const isExist = BlockRepository.isExist(block.id);
-        SyncRepository.sendCommonBlocksExist({ isExist }, peer, requestId);
+        const peerNetwork = PeerNetworkRepository.get(peerAddress);
```
Comment (macaqueee) [inline core/service/sync.ts]: `networkPeer`

File: core/util/socket.ts [hunk 17]
```diff
@@ -0,0 +1,9 @@
+export const sortFunction = (a: { id: any }, b: { id: any }) => {
```
Comment (macaqueee) [inline core/util/socket.ts]: name it generic

File: core/service/peer.ts (hunk 10 above)
Comment (bogdan34) [reply core/service/peer.ts]: +

File: core/controller/block.ts [hunk 18]
```diff
@@ -21,8 +21,9 @@ interface BlockGenerateRequest {
 
 class BlockController extends BaseController {
 
-    @MAIN(ActionTypes.BLOCK_RECEIVE)
+   @MAIN(ActionTypes.BLOCK_RECEIVE)
     public async onReceiveBlock({ data }: { data: { block: BlockModel } }): Promise<ResponseEntity<void>> {
+        logger.debug(`[Controller][Block][onReceiveBlock] height: ${data.block.height} relay: ${data.block.relay}`);
```
Comment (axelsheva) [inline core/controller/block.ts]: remove it, it will be flooding

**Pull request merged**